packages/core/watcher/src/Watcher.js

```diff
@@ -1,27 +1,161 @@
-const FSWatcher = require('fswatcher-child');
+const fork = require('child_process').fork;
+const optionsTransfer = require('./options');
 const Path = require('path');
+const {EventEmitter} = require('events');
+const {errorUtils} = require('@parcel/utils');
 
 /**
  * This watcher wraps chokidar so that we watch directories rather than individual files on macOS.
  * This prevents us from hitting EMFILE errors when running out of file descriptors.
  * Chokidar does not have support for watching directories on non-macOS platforms, so we disable
  * this behavior in order to prevent watching more individual files than necessary (e.g. node_modules).
  */
-class Watcher {
-  constructor() {
-    // FS events on macOS are flakey in the tests, which write lots of files very quickly
-    // See https://github.com/paulmillr/chokidar/issues/612
-    this.shouldWatchDirs =
-      process.platform === 'darwin' && process.env.NODE_ENV !== 'test';
-    this.watcher = new FSWatcher({
-      useFsEvents: this.shouldWatchDirs,
+class Watcher extends EventEmitter {
+  constructor(
+    options = {
+      // FS events on macOS are flakey in the tests, which write lots of files very quickly
+      // See https://github.com/paulmillr/chokidar/issues/612
+      useFsEvents:
+        process.platform === 'darwin' && process.env.NODE_ENV !== 'test',
       ignoreInitial: true,
       ignorePermissionErrors: true,
       ignored: /\.cache|\.git/
-    });
-
+    }
+  ) {
+    super();
+    this.options = optionsTransfer.encode(options);
+    this.watchedPaths = new Set();
+    this.child = null;
+    this.ready = false;
+    this.readyQueue = [];
     this.watchedDirectories = new Map();
     this.stopped = false;
+
+    this.on('ready', () => {
+      this.ready = true;
+      for (let func of this.readyQueue) {
+        func();
+      }
+      this.readyQueue = [];
+    });
+
+    this.startchild();
+  }
+
+  startchild() {
```

Comment:
wonder if it would be worth it to run this inside the existing parcel worker farm at some point. may be worth experimenting with?

Reply:
Yeah probably a good idea